allennlp/data/fields/text_field.py

```diff
@@ -114,17 +145,18 @@ def as_tensor(self,
             # than a LongTensor here, and it's not clear how to signal that.  Maybe we'll need to
             # add a class method to TokenIndexer to tell us the type?  But we can worry about that
             # when there's a compelling use case for it.
-            tensor = torch.LongTensor(padded_array)
-            tensors[indexer_name] = tensor if cuda_device == -1 else tensor.cuda(cuda_device)
+            indexer_tensors = {key: torch.LongTensor(array) for key, array in padded_array.items()}
+            if cuda_device > -1:
+                for key in indexer_tensors.keys():
+                    indexer_tensors[key] = indexer_tensors[key].cuda(cuda_device)
+            tensors.update(indexer_tensors)
         return tensors
 
     @overrides
-    def empty_field(self):
+    def empty_field(self, vocab: Vocabulary):
         # pylint: disable=protected-access
         text_field = TextField([], self._token_indexers)
-        # This needs to be a dict of empty lists for each token_indexer,
-        # for padding reasons in ListField.
-        text_field._indexed_tokens = {name: [] for name in self._token_indexers.keys()}
+        text_field.index(vocab)
```

Comment:
This is a bit of a kludge to pass the vocabulary as an argument just so we can index the text field here. The reason you need to do this is because you don't know how many empty names to add, right? Another possible way to do this would be to add a `get_keys(index_name: str) -> List[str]` method to the token indexers, so that you could do:
```
array_keys = [x for name, indexer in self._token_indexers.items() for x in indexer.get_keys(name)]
text_field._indexed_tokens = {name: [] for name in array_keys}
```

The default implementation for all token_indexers could return the name directly, and could be overridden for an indexer which returns multiple arrays. Can you see anything wrong with that? It would remove the necessity to pass a vocab object to all of the empty_fields, which is nice.

Reply:
How about we just pass the vocab to the token indexer constructor?  Then we can eliminate it from all the various methods (seems we pass it in several different places that could all be eliminated).  Is there a reason this won't work?

Reply:
another option (assuming that you'd index a text field before ever calling empty_field) would be for `index(self, vocab)` to hang onto the vocab (presumably you wouldn't index the same text field with multiple vocabs), and then empty_field could just use that?

Reply:
I don't think we can make the assumption that  `empty_field` is always called after` index`?

Reply:
Refactored the code to remove `vocab` as an argument to `empty_field` and added `get_keys` to the interface for `TokenIndexer`.